### PR TITLE
FORMS-821: Used OS2Web datalookup for CPR and CVR data

### PR DESCRIPTION
**Pull request by rimi-itk**
https://jira.itkdev.dk/browse/FORMS-821

* Uses [OS2Web datalookup](https://github.com/OS2web/os2web_datalookup) for CPR and CVR data.
* Adds handling of physical post (“forsendelse”).

### CODE REVIEW

File: src/Commands/CommandsSF1601.php [hunk 1]
```diff
@@ -155,7 +205,24 @@ private function dump(string $dump, Message $message) {
...
+  private function forsendelseDump(string $dump, ForsendelseI $forsendelse) {
+    switch ($dump) {
+      case self::HIDDEN_CONTENT:
+        break;
+
+      default:
+        throw new InvalidOptionException(sprintf(
+          'Invalid dump option %s. Must be one of %s',
+          json_encode($dump),
+          implode(', ', array_map('json_encode', self::MEMO_DUMP_OPTION_VALUES))
```
Comment (jekuaitk) [inline src/Commands/CommandsSF1601.php]: This doesn't seem to do anything interesting, but im guessing it's gonna be expanded upon later?

File: src/Commands/CommandsSF1601.php [hunk 2]
```diff
@@ -177,7 +244,7 @@ private function getData(int $submissionId, string $handlerId, array $options) {
       catch (\JsonException $exception) {
         throw new InvalidArgumentException(sprintf('Submission data must be valid JSON.'));
       }
-      if (!is_array($submissionData) || !$this->isAssoc($submissionData)) {
+      if (!is_array($submissionData) || array_is_list($submissionData)) {
         throw new InvalidArgumentException(sprintf('Submission data must be an associative array.'));
```
Comment (jekuaitk) [inline src/Commands/CommandsSF1601.php:261]: Should this exception message be updated accordingly?

File: src/Commands/CommandsSF1601.php [hunk 3]
```diff
@@ -191,7 +258,7 @@ private function getData(int $submissionId, string $handlerId, array $options) {
       catch (\JsonException $exception) {
         throw new InvalidArgumentException(sprintf('Handler configuration must be valid JSON.'));
       }
-      if (!is_array($settings) || !$this->isAssoc($settings)) {
+      if (!is_array($settings) || array_is_list($settings)) {
         throw new InvalidArgumentException(sprintf('Handler configuration must be an associative array.'));
```
Comment (jekuaitk) [inline src/Commands/CommandsSF1601.php:275]: Should this exception message be updated accordingly?

File: src/Commands/CommandsSF1601.php (hunk 1 above)
Comment (rimi-itk) [reply src/Commands/CommandsSF1601.php]: Maybe … But I don't think so … https://github.com/itk-dev/os2forms_digital_post/pull/37/commits/9b0bd69bad32fb5ddba5eabc92d9d9d5d11c5bc4#diff-277313824881e121c257dfac232862fef0d2320ab160c36610bdda1d75c27b88R225

File: src/Commands/CommandsSF1601.php (hunk 2 above)
Comment (rimi-itk) [reply src/Commands/CommandsSF1601.php:261]: No the message is still valid.

If `$submissionData` is not an array then it's not an associative array. And if it's a list (and an array) the it's not an associative array.

File: src/Commands/CommandsSF1601.php (hunk 3 above)
Comment (rimi-itk) [reply src/Commands/CommandsSF1601.php:275]: No the message is still valid.

**Review by jekuaitk — APPROVED** [review]

File: composer.json [hunk 4]
```diff
@@ -34,15 +18,17 @@
     "minimum-stability": "dev",
     "prefer-stable": true,
     "require": {
+        "php": "^8.1",
         "ext-soap": "*",
         "dompdf/dompdf": "^2.0",
         "drupal/webform": "^6.0",
+        "drush/drush": "^10 || ^11",
```
Comment (jekuaitk) [inline composer.json]: Could we add `^12`?

File: src/Helper/ForsendelseHelper.php [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  /**
+   * Create modtager.
+   */
+  private function createModtager(CprLookupResult|CompanyLookupResult|null $recipient): ForsendelseModtager {
+    $afsendelseModtager = new AfsendelseModtager();
+    $modtagerAdresse = (new ModtagerAdresse());
+
+    if ($recipient instanceof CprLookupResult) {
+      // @see https://digitaliseringskataloget.srvitkhulk.itkdev.dk/digitaliseringskataloget.dk/sf1601/SF1601%20Bilag%2020211025/SF1601%20Postkomponent%20-%20KombiPostAfsend%20-%20Feltbeskrivelse.pdf#page=7
+      $afsendelseModtager->setCPRnummerIdentifikator('0000000000');
```
Comment (jekuaitk) [inline src/Helper/ForsendelseHelper.php:88]: 😆

**Review by jekuaitk — APPROVED** [review]

**Review by jekuaitk — APPROVED** [review]
🥳

**Review by jekuaitk — APPROVED** [review]

**Pull request merged**